py/embedded-server/deephaven_server/__init__.py

```diff
@@ -3,14 +3,93 @@
 from .server import Server
 
 from deephaven_internal.jvm import check_py_env
+
 check_py_env()
 
 # Note: this is the _distribution_ name, not the _package_ name. Until 3.10, there is not an easy way to get the
 # distribution name from the package name.
 # https://docs.python.org/3/library/importlib.metadata.html#package-distributions
-__version__ = importlib.metadata.version('deephaven-server')
+__version__ = importlib.metadata.version("deephaven-server")
+
+
+@click.group()
+def cli():
+    """
+    Command-line interface entry point for the Deephaven embedded server application.
+    Accepts a command to start the application and the arguments to use.
+    """
+    pass
+
+
+@cli.command()
+@click.option("--host", default=None, help="The host to bind to.")
+@click.option("--port", default=None, type=int, help="The port to bind to.")
+@click.option(
+    "--key",
+    default=None,
+    help="Start the server in pre-shared key mode. If the key is empty, a random key will be used. Cannot be used if anonymous is specified.",
+)
+@click.option(
+    "--anonymous",
+    is_flag=True,
+    help="Start the server in anonymous mode. Cannot be used if key is specified.",
+)
+@click.option("--jvm-args", default=None, help="The JVM arguments to use.")
+def server(host, port, key, anonymous, jvm_args):
+    """
+    Start the Deephaven server.
+    """
+    click.echo("Starting Deephaven server...")
+
+    if anonymous and key is not None:
+        raise click.ClickException("Cannot specify both --anonymous and --key")
+
+    if jvm_args is None:
+        jvm_args = ""
+    jvm_args = jvm_args.split()
+
+    if anonymous:
+        jvm_args += [f"-DAuthHandlers=io.deephaven.auth.AnonymousAuthenticationHandler"]
+    elif key is not None:
+        jvm_args += [
+            f"-DAuthHandlers=io.deephaven.authentication.psk.PskAuthenticationHandler"
+        ]
+        # We are relying on server to set random key otherwise
+        if key:
+            jvm_args += [f"-Dauthentication.psk={key}"]
+
+    s = Server(host=host, port=port, jvm_args=jvm_args)
+    s.start()
+
+    target_url_or_default = s.server_config.target_url_or_default
+    authentication_type = None
+    authentication_url = target_url_or_default
+    for authentication_handler in s.authentication_handlers:
+        authentication_urls = authentication_handler.urls(target_url_or_default)
+        if authentication_urls:
+            authentication_type = authentication_handler.auth_type
+            authentication_url = authentication_urls[0]
+            break
+
+    webbrowser.open(authentication_url)
+
+    click.echo(
+        f"Deephaven is running at {authentication_url} with authentication type {authentication_type}"
+    )
+    click.echo("Press Control-C to exit")
+
+    def signal_handler(sig, frame):
+        click.echo("Exiting Deephaven...")
+        sys.exit(0)
+
+    signal.signal(signal.SIGINT, signal_handler)
+    signal.pause()
```

Comment:
This is looking great.

1. I would add an optional argument to `server` that takes a path to a script file to run before getting into the wait loop.  Basically a way to populate DH with a quick command.
2. I would add a new `run` command to run a script.  Basically the same as `server` without the auth, browser, and wait.  Batch mode.

Reply:
Both of those options sound great. Mind if it's a future PR?:
1. That's kind of like application mode, would we want to specify the "app" to start with? I believe the only way to do that now is by setting the app.d directory, so we'd want to think how we have this play out/work with application mode.
2. Similar to question 1, do we still have the application mode scripts run or just the file you specified? What config do we want to override in that case?

In any case with the console scripts added, we can add new commands/options fairly easily.

Reply:
Let's do it soon.